middleware/ratelimiter/ratelimiter.go

```diff
@@ -0,0 +1,88 @@
+package ratelimiter
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/System-Glitch/goyave/v3"
+)
+
+// LimiterConfig for setting configuration for the limiter middleware
+type LimiterConfig struct {
+	// Maximum number of requests in a client can send
+	RequestQuota int
+
+	// Duration or time taken until the quota expires and renews
+	QuotaDuration time.Duration
+
+	// Unique identifier for requestors. Can be userID or IP
+	// Defaults to Remote Address if it is empty
+	ClientID string
+
+	// Handles response when rate limit exceeds
+	ResponseHandler goyave.Handler
+}
+
+// New initializes new a rate limiter middleware
+func New(configFn func(request *goyave.Request) LimiterConfig) goyave.Middleware {
+
+	lstore := newLimiterStore()
+
+	return func(next goyave.Handler) goyave.Handler {
+
+		return func(response *goyave.Response, request *goyave.Request) {
+
+			config := configFn(request)
+
+			if config.RequestQuota == 0 || config.QuotaDuration == 0 {
+				next(response, request)
+				return
+			}
+
+			if config.ResponseHandler == nil {
+				config.ResponseHandler = defaultResponseHandler
+			}
+
+			if config.ClientID == "" {
+				config.ClientID = defaultClientID(request)
+			}
+
+			key := config.ClientID
+
+			l := lstore.get(key)
+
+			if l == nil {
+				l = newLimiter(config.RequestQuota, config.QuotaDuration)
+				lstore.set(key, l)
+			}
+
+			if time.Since(l.updatedAt) < l.quotaDuration {
```

Comment:
There is a possible race condition here. The `l` variable is not protected by mutex. The call through `lstore` is, though, but you should move that logic to the `limiter` struct.

As it is a pointer, make your changes directly to the `l` variable and don't use `lstore` as a proxy.

Reply:
I am using mutex locks on `lstore` functions such as `set, increments and reset` and none in the `limiter` functions. Making changes directly to `l` variable (without mutex locks), wouldn't it cause race conditions?

Sorry, my knowledge in this area is limited.

Reply:
Imagine two requests for the same user occur on two goroutines.

Goroutine 1 locks in `lstore.reset()`, and changes the updateAt value. At the same time, goroutine 2 checks if `time.Since(l.updatedAt) < l.quotaDuration`, but without locking anything. There will be a race condition between goroutine 1 (write) and goroutine 2 (read).

Reply:
Oh ok. I understand now. I will look into this